src/SafeProvider.ts

```diff
@@ -12,22 +32,64 @@ export type SafeProviderProps = {
 const queryClient = new QueryClient()
 
 export function SafeProvider(params: React.PropsWithChildren<SafeProviderProps>) {
-  const { children, config } = params
+  const [initialized, setInitialized] = useState(false)
+  const [config, setConfig] = useState(params.config)
+  const [publicClient, setPublicClient] = useState<SafeClient>()
+  const [signerClient, setSignerClient] = useState<SafeClient>()
+
+  const wagmiConfig = useMemo(
+    () =>
+      createConfig({
+        chains: [config.chain],
+        transports: { [config.chain.id]: config.transport }
+      }),
+    [config.chain, config.transport]
+  )
+
+  useEffect(() => {
+    Promise.all([
+      createPublicClient(config).then(setPublicClient),
+      isSafeConfigWithSigner(config) ? setSigner(config.signer) : Promise.resolve()
+    ])
+      .then(() => {
+        setInitialized(true)
+      })
+      .catch((err) => {
+        throw new InitializeSafeProviderError('Failed to initialize clients.', err)
+      })
+  }, [config])
+
+  const setSigner = useCallback(
+    async (signer: string | undefined) => {
+      if (signer) {
+        try {
+          const newSignerClient = await createSignerClient({ ...config, signer })
+          setSignerClient(newSignerClient)
+        } catch (err) {
+          throw new InitializeSafeProviderError('Failed to initialize signer client.', err)
+        }
+      } else if (signerClient) {
+        setSignerClient(undefined)
+      }
+    },
+    [config, signerClient]
+  )
 
-  const wagmiConfig = createConfig({
-    chains: [config.chain],
-    transports: { [config.chain.id]: config.transport }
-  })
+  if (!config.provider) {
+    throw new InitializeSafeProviderError('Provider not set in config.')
+  }
 
-  const props = { value: config }
+  const props = {
+    value: { initialized, config, setConfig, setSigner, publicClient, signerClient }
+  }
 
   return createElement(
     SafeContext.Provider,
     props,
     createElement(
```

Comment:
Why are you using the createElement functions instead .tsx files with proper tag syntax?

Reply:
I used it to avoid configuring JSX support for jest. Had some trouble when trying to configure it, but will check it again